codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,329 +16,506 @@
+    private final PropertyKey streamKey;
 
-    private final Http2ConnectionHandler http2Handler;
-    private final boolean server;
+    private final Integer initialFlowControlWindowSize;
 
     private ChannelHandlerContext ctx;
-    private ChannelHandlerContext http2HandlerCtx;
 
-    /**
-     * Construct a new handler.
-     *
-     * @param server {@code true} this is a server
-     */
+    /** Number of buffered streams if the {@link StreamBufferingEncoder} is used. **/
+    private int numBufferedStreams;
+
     public Http2FrameCodec(boolean server) {
-        this(server, HTTP2_FRAME_LOGGER);
+        this(server, null);
     }
 
-    /**
-     * Construct a new handler.
-     *
-     * @param server {@code true} this is a server
-     */
     public Http2FrameCodec(boolean server, Http2FrameLogger frameLogger) {
         this(server, new DefaultHttp2FrameWriter(), frameLogger, Http2Settings.defaultSettings());
     }
 
-    // Visible for testing
-    Http2FrameCodec(boolean server, Http2FrameWriter frameWriter, Http2FrameLogger frameLogger,
-                    Http2Settings initialSettings) {
-        // TODO(scott): configure maxReservedStreams when API is more finalized.
-        Http2Connection connection = new DefaultHttp2Connection(server);
-        frameWriter = new Http2OutboundFrameLogger(frameWriter, frameLogger);
-        Http2ConnectionEncoder encoder = new DefaultHttp2ConnectionEncoder(connection, frameWriter);
-        Long maxHeaderListSize = initialSettings.maxHeaderListSize();
-        Http2FrameReader frameReader = new DefaultHttp2FrameReader(maxHeaderListSize == null ?
-                new DefaultHttp2HeadersDecoder(true) :
-                new DefaultHttp2HeadersDecoder(true, maxHeaderListSize));
-        Http2FrameReader reader = new Http2InboundFrameLogger(frameReader, frameLogger);
-        Http2ConnectionDecoder decoder = new DefaultHttp2ConnectionDecoder(connection, encoder, reader);
+    Http2FrameCodec(Http2ConnectionEncoder encoder, Http2ConnectionDecoder decoder, Http2Settings initialSettings,
+                    long gracefulShutdownTimeoutMillis) {
+        super(decoder, encoder, initialSettings);
+
         decoder.frameListener(new FrameListener());
-        http2Handler = new InternalHttp2ConnectionHandler(decoder, encoder, initialSettings);
-        http2Handler.connection().addListener(new ConnectionListener());
-        this.server = server;
+        connection().addListener(new ConnectionListener());
+        connection().remote().flowController().listener(new Http2RemoteFlowControllerListener());
+        streamKey = connection().newKey();
+        initialFlowControlWindowSize = initialSettings.initialWindowSize();
+        gracefulShutdownTimeoutMillis(gracefulShutdownTimeoutMillis);
     }
 
-    Http2ConnectionHandler connectionHandler() {
-        return http2Handler;
+    private Http2FrameCodec(boolean server, Http2FrameWriter frameWriter, Http2FrameLogger frameLogger,
```

Comment:
https://github.com/netty/netty/pull/6896#discussion_r132096300

> can we funnel this method through a single constructor, or share this initialization code? It is not clear why `gracefulShutdownTimeoutMillis` isn't set here. In general the initialization can change and its nice to have this consolidated so we don't miss things in the future.

> Can you clarify why we are moving away from `Http2FrameCodecBuilder`? It would be nice to keep the logic consolidated instead of duplication in here and `Http2ConnectionHandler`.